includes/class-wc-gateway-stripe.php

```diff
@@ -1266,4 +1267,70 @@ public function settings_api_sanitized_fields( $settings ) {
 		}
 		return $settings;
 	}
+
+	/**
+	 * Get required setting keys for setup.
+	 *
+	 * @return array Array of setting keys used for setup.
+	 */
+	public function get_required_settings_keys() {
+		return array( 'publishable_key', 'secret_key' );
+	}
+
+	/**
+	 * Get the connection URL.
+	 *
+	 * @return string Connection URL.
+	 */
+	public function get_connection_url( $return_url = '' ) {
+		$api     = new WC_Stripe_Connect_API();
+		$connect = new WC_Stripe_Connect( $api );
+
+		$url = $connect->get_oauth_url( $return_url );
+
+		return is_wp_error( $url ) ? null : $url;
+	}
+
+	/**
+	 * Get help text to display during quick setup.
+	 *
+	 * @return string
+	 */
+	public function get_setup_help_text() {
+		return sprintf(
+			__( 'Your API details can be obtained from your <a href="%s">Stripe account</a>. Don’t have a Stripe account? <a href="%s">Create one.</a>', 'woocommerce-gateway-stripe' ),
+			'https://stripe.com/docs/keys',
+			'https://dashboard.stripe.com/register'
+		);
+	}
+
+	/**
+	 * Determine if the gateway still requires setup.
+	 *
+	 * @return bool
+	 */
+	public function needs_setup() {
+		return ! $this->get_option( 'publishable_key' ) || ! $this->get_option( 'secret_key' );
+	}
+
+	/**
+	 * Updates the test mode based on keys provided when setting up the gateway via onboarding.
+	 *
+	 * @return array
+	 */
+	public function update_onboarding_settings( $settings ) {
+		if ( ! empty( $settings['publishable_key'] ) && ! empty( $settings['secret_key'] ) ) {
+			if ( strpos( $settings['publishable_key'], 'test_' ) === 0 || strpos( $settings['secret_key'], 'test_' ) === 0 ) {
+				$settings['test_publishable_key'] = $settings['publishable_key'];
+				$settings['test_secret_key'] = $settings['secret_key'];
+				unset( $settings['publishable_key'] );
+				unset( $settings['secret_key'] );
+				$settings['testmode'] = 'yes';
+			} else {
+				$settings['testmode'] = 'no';
```

Comment:
Would this not turn off test mode any time the gateway is updated after live keys are set (via REST API, which looks like the only time this filter is run)?

Reply:
It does.  Can both live keys and test keys exist simultaneously?

This would be much better if we could check if the request was being made from WooCommerce Admin, but AFAIK the only params allowed are the predefined settings in the gateway.

Reply:
Sorry, didn't see this comment before replying below – yup, live and test keys can both be set, and it's just the `testmode` setting that switches between them.